tools/compactsnoop.py

```diff
@@ -0,0 +1,389 @@
+#!/usr/bin/python
+# @lint-avoid-python-3-compatibility-imports
+#
+# compactsnoop  Trace compact zone and print details including issuing PID.
+#       For Linux, uses BCC, eBPF.
+#
+# This uses in-kernel eBPF maps to cache process details (PID and comm) by
+# compact zone begin, as well as a starting timestamp for calculating
+# latency.
+#
+# Copyright (c) 2019 Ethercflow
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 11-NOV-2019   Ethercflow   Created this.
+
+from __future__ import print_function
+from bcc import BPF
+import argparse
+from datetime import datetime, timedelta
+
+# arguments
+examples = """examples:
+    ./compactsnoop          # trace all compact stall
+    ./compactsnoop -T       # include timestamps
+    ./compactsnoop -d 10    # trace for 10 seconds only
+    ./compactsnoop -K       # output kernel stack trace
+    ./compactsnoop -e       # show extended fields
+"""
+
+parser = argparse.ArgumentParser(
+    description="Trace compact zone",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples,
+)
+parser.add_argument("-T", "--timestamp", action="store_true",
+        help="include timestamp on output")
+parser.add_argument("-p", "--pid", help="trace this PID only")
+parser.add_argument("-d", "--duration",
+        help="total duration of trace in seconds")
+parser.add_argument("-K", "--kernel-stack", action="store_true",
+        help="output kernel stack trace")
+parser.add_argument("-e", "--extended_fields", action="store_true",
+        help="show system memory state")
+parser.add_argument("--ebpf", action="store_true", help=argparse.SUPPRESS)
+args = parser.parse_args()
+debug = 0
+if args.duration:
+    args.duration = timedelta(seconds=int(args.duration))
+
+NO_EXTENDED = """
+#ifdef EXTNEDED_FIELDS
+#undef EXTNEDED_FIELDS
+#endif
+"""
+
+EXTENDED = """
+#define EXTNEDED_FIELDS    1
+"""
+
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/sched.h>
+#include <linux/mmzone.h>
+#include <linux/compaction.h>
+
+struct val_t {
+    int nid;
+    int idx;
+    int order;
+    int sync;
+#ifdef EXTNEDED_FIELDS
+    int fragindex;
+    int low;
+    int min;
+    int high;
+    int free;
+#endif
+    u64 ts;    // compation begin time
+};
+
+struct data_t {
+    u32 pid;
+    u32 tid;
+    int nid;
+    int idx;
+    int order;
+    u64 delta;
+    u64 ts;    // compation end time
+    int sync;
+#ifdef EXTNEDED_FIELDS
+    int fragindex;
+    int low;
+    int min;
+    int high;
+    int free;
+#endif
+    int status;
+    int stack_id;
+    char comm[TASK_COMM_LEN];
+};
+
+BPF_HASH(start, u64, struct val_t);
+BPF_PERF_OUTPUT(events);
+BPF_STACK_TRACE(stack_traces, 10240);
```

Comment:
inside the kernel, the number of entries will be rounded up to power of 2.
https://github.com/torvalds/linux/blob/master/kernel/bpf/stackmap.c#L114
could you just round up to power of 2 here to avoid confusion.

Reply:
Thanks a lot for give me this details, I'll fix this.